pkg/cri/server/sandbox_run.go

```diff
@@ -377,6 +403,84 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		return nil, fmt.Errorf("failed to start sandbox container task %q: %w", id, err)
 	}
 
+	// If userns is enabled, then the netns was created by the OCI runtime
+	// when starting the sandbox. The OCI runtime needs to create the netns
+	// because, if userns is in use, the netns needs to be owned by the
+	// userns. So, let the OCI runtime just handle this for us.
+	// If the netns is not owned by the userns several problems will happen.
+	// For instance, the container will lack permission (even if
+	// capabilities are present) to modify the netns or, even worse, the OCI
+	// runtime will fail to mount sysfs:
+	// 	https://github.com/torvalds/linux/commit/7dc5dbc879bd0779924b5132a48b731a0bc04a1e#diff-4839664cd0c8eab716e064323c7cd71fR1164
+	if podNetwork && userNsEnabled {
```

Comment:
IMO, we should setup CNI before `task.Start`, which looks like that it is runC-Prestart hook. If there is any restart-containerd action, the cri plugin will cleanup `CREATED` status during recovery. It will be safe.

Reply:
I was actually hoping we'd be able to also move the !userNS network setup down and/or if possible try to make it parallel for performance reasons.. context: https://github.com/containerd/containerd/pull/5904

Reply:
doesn't the userns get created in runc create.. up in NewTask() before we call task start?

Reply:
The runC-create will create namespace. So after NewTask we can mount the netns namespace before task.Start?

Reply:
oh, right. Great catch, I've moved it just now to after NewTask and before start, it all seems to work fine. Will be reflected here in the next push, thanks!

Reply:
Should be solved now

Reply:
Sorry I don't think it is correct. 😂 

With your original patch, it is like 

```
1. skip cni.setup if it enables userns
2. NewTask
3. task.Start
4. cni.Setup if it enables userns
```

But now. It is 

```
1. NewTask
2. cni.Setup
3. task.Start
```

It breaks current sandbox-liked runtime usage. The sandbox-liked runtime needs pre-allocated netns for NewTask.

So I think the steps should be like

```
1. skip cni.setup if it enables userns
2. NewTask
3. cni.setup if enables userns
4. task.Start.
```

Reply:
Oh, sorry. I made it like that now.

Why would the previous version break sandbox-like runtimes? Don't they create the netns as the OCI spec says?

I see it can break windows, as I didn't found a way to join a netns created by another process in the godoc (maybe there is?). So that alone is enough to keep both cases.

Reply:
In some cases from kata-runtime, it needs pre-allocated netns with cniSetup so that the shim can init kvm with networking interface. I also remember that gVisor also works in this way. So we should keep it compatible. 

> as I didn't found a way to join a netns created by another process in the godoc (maybe there is?)

maybe unshare? https://github.com/containerd/containerd/pull/7513

> So that alone is enough to keep both cases.

Yeah. It looks good to me. let's move it forward.